### ISSUE TITLE
【質問】【第3章：p.173-174】3-7_PSPNet_training の loss について

### ISSUE THREAD

**Issue by mitgrad2018** [YutaroOgawa/pytorch_advanced#98]
素晴らしい本を執筆してくださり大変感謝しております。
勉強しているなかで、ちゃんと理解できていない点があり、質問させていただきたく思います。

質問(1)：
loss.item()はミニバッチで平均した損失であると理解しておりますが、
下記コードにおいて、さらにbatch_sizeで除算しているのはなぜでしょうか。
loss.item() / batch_size * batch_multiplier

質問(2)：
下記コードにて、batch_multiplierで除算し、
loss = criterion(outputs, anno_class_images.long()) / batch_multiplier
その後、下記コードにて、batch_multiplierを掛け戻しているようですが、敢えて除算して掛け戻す操作にしているのはなぜでしょうか。
epoch_train_loss += loss.item() * batch_multiplier

お忙しいところ恐れ入りますが、ご回答いただけますと幸甚です。

**Comment by YutaroOgawa** [comment]
@mitgrad2018 さま

ご質問ありがとうございます。返答が遅くなり申し訳ございません。

非常に丁寧にお読みいただき、ありがとうございます。

質問(1)：
loss.item() / batch_size * batch_multiplier

で batch_size で割る必要はあるのか？

p.169のlossの定義で

```python
loss = F.cross_entropy(outputs[0], targets, reduction='mean')
loss_aux = F.cross_entropy(outputs[1], targets, reduction='mean')
return loss+self.aux_weight*loss_aux
```
と記載しており、すでにミニバッチサイズで割り算されています。

そのため、 @mitgrad2018 さまのご指摘通り、batch_sizeで割るのは不自然であり、

p. 173-174
（誤り）
loss.item() / batch_size * batch_multiplier

↓
（正）
loss.item()  * batch_multiplier

となります。こちらを直すと、図3.7.2も直す必要があり、ちょっと大変なので、書籍への反映は少々お待ちください。

ただ、本Issueに誤植ラベルを貼り、他の読者の方々の参考にさせてください。


質問(2)：

>下記コードにて、batch_multiplierで除算し、
loss = criterion(outputs, anno_class_images.long()) / batch_multiplier
その後、下記コードにて、batch_multiplierを掛け戻しているようですが、敢えて除算して掛け戻す操作にしているのはなぜでしょうか。
epoch_train_loss += loss.item() * batch_multiplier

これはlossをバックプロパゲーションするので、batch_multiplierを今は3ですが、batch_multiplierを4に変更したりしても、lossへの影響をなくし、学習率を変化させずに済ませるためになります。

以上、どうぞよろしくお願い致します。

**Comment by mitgrad2018** [comment]
丁寧にご回答いただきありがとうございます。

上記質問(2)についてですが、下記の理解で正しいでしょうか。

multiple minibatch内は、backwardが行われる度に勾配が蓄積されるため、
backward を batch_multiplier回(今回の例では3回)実施後は、勾配はbatch_multiplierで除算された勾配の合計(つまりmultiple minibatch内の平均)となり、その勾配に対してoptimizer.stepを実行している。

**Comment by YutaroOgawa** [comment]
@mitgrad2018 さま

はい、全く以下、その通りでございます

>multiple minibatch内は、backwardが行われる度に勾配が蓄積されるため、
backward を batch_multiplier回(今回の例では3回)実施後は、勾配はbatch_multiplierで除算された勾配の合計(つまりmultiple minibatch内の平均)となり、その勾配に対してoptimizer.stepを実行している。

そして、batch_multiplier=3が4に変化しても、対応するようにコードを実装していることになります。


重要なご質問、ならびに理解の確認をいただき、誠にありがとうございます。
（私も自分の説明がきちんと多くの読者に理解してもらえているのか不明なので、このように理解の確認をいただけることは、非常に助かります）

**Comment by mitgrad2018** [comment]
小川様

早速のご確認ありがとうございます。
このような機会を設けていただき大変感謝しております。
他の質問も別途させていただいており恐縮ですが、またご教示いただけますと幸甚です。

**Comment by YutaroOgawa** [comment]
@mitgrad2018 さま

ご質問ありがとうございます。
本書の対応に使える時間を確保できず、きちんと確認しきれないのですが、
いただいた内容で、まったく問題なく、良いと思います。

本書、lossの単位がバラバラなので、1サンプル単位になるように丁寧にそろえていただき、ありがとうございます！

**Comment by mitgrad2018** [comment]
小川様

早速ご回答いただきありがとうございます。
お忙しいところご確認いただき感謝いたします。
